AP_LeakDetector: add manual leak-pin selection to extend board compatibility

paulblum opened the pull request:
This pull request addresses issue #22013.

At present, the AP_LeakDetector library used in ArduSub contains compilation directives that only allow this sensor to initialize on FMUv3, FMUv5, or Linux navigator boards. Sensor initialization is neglected for all other board types, without providing any sort of notice to the user.

The changes proposed will here remove the initialization's dependence on board type. In this updated code, leak detectors can be initialized over connection to arbitrary pin number on any type of board. A new advanced parameter, `LEAKx_TYPE`, has been introduced that can be used to enable manual configuration of the leak detector pin. In order to use this feature, a user would set the `LEAKx_TYPE` parameter to indicate the signal type (analog or digital) of an appropriately configured pin on their board. The number of this pin should then be set as the value of the existing `LEAKx_PIN` parameter.

ES-Alexander commented:
@Williangalvani

---

@paulblum The generalisation here has merit - supporting all boards is clearly superior to a limited subset if it works well. That said, the proposed required parameter is somewhat undesirable, given it puts additional responsibility onto the user (or at least their configuration software) to know about it, and to know the pin numbers and capabilities of their board. It might not be possible to add general support without that though, so 🤷‍♂️ 

For some context (as I understand it):
- the existing board dependence is because ArduSub has only been actively tested on those board types, so those are the only ones that are actively supported, and
- the existing pin dependence is to avoid the user being able to accidentally configure pins that are not capable of being used as analog/digital inputs (as applicable given the board type)

paulblum commented:
Thanks for your feedback @ES-Alexander!

I just pushed a new commit that restores the original behavior of this code while adding manual configuration as a new (optional) advanced parameter named `LEAKx_TYPE`.

With this version, a user would be able to use `LEAKx_TYPE` to indicate the signal type (analog/digital) of any appropriately configured pin on their board; the number of this pin would then be set as the value of the existing `LEAKx_PIN` parameter.

Let me know what you think :smile:

ES-Alexander commented:
@Williangalvani the revised version seems like a decent compromise of improved usability vs our concerns of backwards compatibility and users being able to accidentally override important functionality:

```python
if known_board():
    ... # do current behaviour, with hardcoded pins
elif user_specified_pin_type():  # untested/unknown board, explicitly opt-in
    ... # register arbitrary pin as specified type (possibly dodgy)
else:
    pass
```

@paulblum My only remaining concern is the potential for confusion with the parameter being optional, and sometimes having no effect. Perhaps it can be documented as `"NOT USED on Pixhawk, Pixhawk 4, and Navigator flight controllers"`, or `IGNORED` or similar?

paulblum commented:
@ES-Alexander how about something like this?
I updated the descriptions of the new `LEAKx_TYPE` parameter & modified the code so that this parameter can't have any affect on actively supported boards.

Williangalvani commented on libraries/AP_LeakDetector/AP_LeakDetector.cpp:
```diff
@@ -12,12 +12,20 @@ const AP_Param::GroupInfo AP_LeakDetector::var_info[] = {
     // @Values: 0:Low,1:High
     // @User: Standard
-    AP_GROUPINFO("1_LOGIC", 2, AP_LeakDetector, _default_reading[0], 0),
+    AP_GROUPINFO("1_LOGIC", 3, AP_LeakDetector, _default_reading[0], 0),
```
I think these parameter id changes can cause issues when upgrading from an older firmware.
Can we keep the old ones as they are and just use new numbers for the new ones?

Williangalvani commented:
This is looking good. I like it!
I'll have someone else do double-check it though.

IamPete1 reviewed (changes requested):
No need to change the existing param IDs, that will result in values getting lost in the upgrade. Just add the new params with 7,8,9. They do not need to me in numerical order in the table. 

With some param conversion we would be able to un-tangle the type by pin stuff that is a little odd.

IamPete1 commented on libraries/AP_LeakDetector/AP_LeakDetector.cpp:
```diff
@@ -85,10 +110,16 @@ void AP_LeakDetector::init()
         default:
-            _drivers[i] = NULL;
-            break;
+            if (_type[i] == 0) {
```
Better to use the switch to set the defualt type. EG:

```
        switch (_pin[i]) {
#if (CONFIG_HAL_BOARD_SUBTYPE == HAL_BOARD_SUBTYPE_CHIBIOS_FMUV3 || \
     CONFIG_HAL_BOARD_SUBTYPE == HAL_BOARD_SUBTYPE_CHIBIOS_FMUV5)
        case 50 ... 55:
            _type[i].set_default(1)
            break;
        case 13 ... 15:
            _type[i].set_default(0)
            break;
#elif CONFIG_HAL_BOARD_SUBTYPE == HAL_BOARD_SUBTYPE_LINUX_NAVIGATOR
        case 27:
            _type[i].set_default(1)
            break;
```
Then you can just switch on type as normal.

paulblum replied:
@IamPete1 is there an advantage to using `set_default()` here instead of just `set()`?

IamPete1 replied:
Set default will only change the param if the user has never set it (its new so they wont). It makes it possible to have a digital on pin 13 if the user configure it that way.

paulblum replied:
Ah so `set_default()` gives users the option to change this param, i.e. providing more freedom on the supported boards rather than locking in their defaults with an overwrite. That makes sense to me!

IamPete1 commented on libraries/AP_LeakDetector/AP_LeakDetector.cpp:
```diff
@@ -85,10 +110,16 @@ void AP_LeakDetector::init()
+            if (_type[i] == 0) {
+                _state[i].instance = i;
+                _drivers[i] = new AP_LeakDetector_Analog(*this, _state[i]);
+            } else if (_type[i] == 1) {
```
Add a enum class for the types.

paulblum commented:
Thank you for the feedback @Williangalvani & @IamPete1, I appreciate the ideas!

I changed the parameter IDs back to their original values & changed initialization to be based on a type switch, which I agree seems much more intuitive.

I also added an enum in the header file that provides names for the set of signal types– is this the best spot/implementation for this?

IamPete1 commented:
Looking good, commits need a squash and prefixing with `AP_LeakDetector:`

paulblum commented:
I think this is all set, but please let me know if you find anything that needs to be modified.

Thanks again for your help @ES-Alexander, @Williangalvani & @IamPete1!

rmackay9 commented on libraries/AP_LeakDetector/AP_LeakDetector.cpp:
```diff
@@ -19,6 +19,14 @@ const AP_Param::GroupInfo AP_LeakDetector::var_info[] = {
 
+    // @Param: 1_TYPE
+    // @DisplayName: Pin type (analog/digital)
+    // @Description: NOT FOR USE with Pixhawk, Pixhawk 4 or Navigator flight controllers. Enables manual leak-pin selection on other boards-- use this parameter to indicate the signal type (0:analog, 1:digital) of an appropriately configured pin, then use the LEAK1_PIN parameter to specify its number.
```
Let's move this, "NOT FOR USE" portion to closer to the end of the parameter description.

rmackay9 commented on libraries/AP_LeakDetector/AP_LeakDetector.cpp:
```diff
@@ -19,6 +19,14 @@ const AP_Param::GroupInfo AP_LeakDetector::var_info[] = {
     AP_GROUPINFO("1_LOGIC", 2, AP_LeakDetector, _default_reading[0], 0),
 
+    // @Param: 1_TYPE
+    // @DisplayName: Pin type (analog/digital)
```
The display name should start with "Leak detector" I think.

rmackay9 commented on libraries/AP_LeakDetector/AP_LeakDetector.h:
```diff
@@ -45,7 +45,12 @@ class AP_LeakDetector {
+        ANALOG=0,
+        DIGITAL=1
+    };
+    AP_Int8 _type[LEAKDETECTOR_MAX_INSTANCES];
```
Could add a comment beside the _type parameter declaration.

tridge commented:
@Williangalvani will test

paulblum commented:
@rmackay9 thank you for the comments– I just pushed a commit that modifies the documentation to incorporate your suggestions, let me know what you think!

Williangalvani commented:
looks good to me, can you squash the commits?

@IamPete1 are you happy with it?

paulblum commented:
Squashed & rebased!

IamPete1 reviewed (approved):
LGTM
